Issue102 adding history

bfitzpat opened the pull request:
Adds a very basic history to the Didact window. It re-opens the files and is not caching the HTML, but does cache the URIs used. Note that the cache is reset to the last didact file opened at workspace closing time if it is available. Otherwise, the history is empty each time the workspace starts. 

To test:
* open a few didact files from the didact view
* use ctrl+shift+left to go back in the history (it loops back from the first item to the last)
* use ctrl+shift+right to go forward in the history (it loops forward from the last item to the first)
* leave a didact file open, close the workspace
* reopen the workspace, add a different didact, the cached didact file should be first in the history

![image](https://user-images.githubusercontent.com/530878/90698339-aeb1ed00-e23d-11ea-904b-0072a797c640.png)

apupier reviewed (commented):
unless it is impossible with current VS Code API, I think we shoudl nto introduce another commands to navigate in history than the default VS Code history

bfitzpat commented:
> unless it is impossible with current VS Code API, I think we shoudl nto introduce another commands to navigate in history than the default VS Code history

I have looked at depth for ways to get to the history service (https://github.com/microsoft/vscode/blob/f74e473238aca7b79c08be761d99a0232838ca4c/src/vs/workbench/services/history/common/history.ts) and had NO good examples of how to integrate with it. 

We can shift to these keys easily:

![image](https://user-images.githubusercontent.com/530878/90770411-fc1a7280-e2ae-11ea-8744-180daa49dc87.png)

We can even probably add a Didact: Clear History command. 

But unless you can point me in the direction of a way to integrate with the built-in VS Code history service, we are going to have to roll our own. Every example I have found has had its own way of capturing and tracking it.

bfitzpat commented:
@apupier  -- I have altered keys to Alt+Up (previous) and Alt+Down (Next) to match the history.showNext/showPrevious key mappings, but won't add simple Up/Down since that would interfere with scrolling in the webview itself.

bfitzpat commented:
Now I have added an explicit clear history command that is user accessible in the palette

bfitzpat commented:
![history-demo-20-AUG-2020](https://user-images.githubusercontent.com/530878/90809216-61398c80-e2de-11ea-95c2-a06131ec540b.gif)

lhein commented on media/main.js:
```diff
@@ -79,36 +90,44 @@ function () {
 				const isAvailable = json.result;
 
+				let requirementMessage = 'Not currently available';
 				let element = document.getElementById(requirementName);
```
element can be made a const and given a type

bfitzpat replied:
const maybe, but see the notes about JS vs. TS in this case, so no type

lhein replied:
my fault

lhein commented on media/main.js:
```diff
@@ -22,6 +22,17 @@ function () {
+	var obj = new Window_History();
+
 	document.body.addEventListener('click', event => {
 		let node = event && event.target;
```
node is not changed any more so could be made a const with a type.

bfitzpat replied:
const maybe, but see the notes about JS vs. TS in this case, so no type

lhein replied:
my fault

lhein commented on media/main.js:
```diff
@@ -79,36 +90,44 @@ function () {
-					const element = links[index];
-					if (element.getAttribute('href')) {
-						const href = element.getAttribute('href');
+					const linkElements = links[index];
```
can we add type here?

bfitzpat replied:
this is a javascript file, not a typescript file, so I cannot add types. it's loaded into the HTML that's generated as part of the didact md/adoc-to-html process along with some stylesheets, so it's not the same as a TS file

lhein replied:
my fault

lhein commented on media/main.js:
```diff
@@ -79,36 +90,44 @@ function () {
-						const href = element.getAttribute('href');
+					const linkElements = links[index];
+					if (linkElements.getAttribute('href')) {
+						const href = linkElements.getAttribute('href');
```
We should try to add types always.

bfitzpat replied:
this is a javascript file, not a typescript file, so I cannot add types. it's loaded into the HTML that's generated as part of the didact md/adoc-to-html process along with some stylesheets, so it's not the same as a TS file

lhein replied:
OMG....I just looked at the code and did not see that :D My fault.

lhein commented on src/didactWebView.ts:
```diff
@@ -31,6 +31,10 @@ export class DidactWebviewPanel {
 	public static currentPanel: DidactWebviewPanel | undefined;
 
 	public static readonly viewType = 'didact';
+	private static readonly didactCachePath = `didact/cache`;
```
can we add types to these ?

bfitzpat replied:
will do types in a separate PR

lhein commented on src/didactWebView.ts:
```diff
@@ -203,6 +208,10 @@ export class DidactWebviewPanel {
 	}
 
+	public setActiveContext(value: boolean) {
+		vscode.commands.executeCommand('setContext', 'didact.webview', value);
```
executeCommand is returning a Thenable means its executed asynchronously. Maybe worth to check if we need to async/await it or not.

bfitzpat replied:
All the examples I saw had it not async -- but it would definitely have some serious ripples if I change it to async. I'm going to leave it unless it becomes a problem.

bfitzpat replied:
I looked into this a bit and it would be a nightmare of ripples -- but even in vscode's own code they're not treating it async - https://github.com/Microsoft/vscode/blob/cd7c7c5fa2c16c6e8281436c35e9a7709cfbd89d/extensions/markdown-language-features/src/features/previewManager.ts#L152

lhein commented on src/didactWebView.ts:
```diff
@@ -406,53 +415,57 @@ export class DidactWebviewPanel {
-			const cachePath = path.join(DidactWebviewPanel.context.globalStoragePath, `didact/cache`);
-			const htmlFilePath = path.join(cachePath, 'currentHtml.html');
-			const titleFilePath = path.join(cachePath, 'currentTitle.txt');
+			const cachePath = path.join(DidactWebviewPanel.context.globalStoragePath, DidactWebviewPanel.didactCachePath);
```
can we add types to these constants?

bfitzpat replied:
will do types in a separate PR

lhein commented on src/didactWebView.ts:
```diff
@@ -406,53 +415,57 @@ export class DidactWebviewPanel {
-				if (fs.existsSync(htmlFilePath)) {
-					let contents = fs.readFileSync(htmlFilePath);
-					return contents.toLocaleString();
+			const cachePath = path.join(DidactWebviewPanel.context.globalStoragePath, DidactWebviewPanel.didactCachePath);
```
can we add type here?

bfitzpat replied:
will do types in a separate PR

lhein commented on src/didactWebView.ts:
```diff
@@ -406,53 +415,57 @@ export class DidactWebviewPanel {
-					return contents.toLocaleString();
+			const cachePath = path.join(DidactWebviewPanel.context.globalStoragePath, DidactWebviewPanel.didactCachePath);
+			if (cachePath) {
+				const filePath = path.join(cachePath, filename);
```
and here

bfitzpat replied:
will do types in a separate PR

lhein commented on src/didactWebView.ts:
```diff
@@ -406,53 +415,57 @@ export class DidactWebviewPanel {
+				const filePath = path.join(cachePath, filename);
+				try {
+					if (fs.existsSync(filePath)) {
+						let contents = fs.readFileSync(filePath);
```
this can be made a const and assign a type

lhein commented on src/doublyLinkedList.ts:
```diff
@@ -0,0 +1,83 @@
+	*/
+	dequeue(): T | undefined {
+		if (this.head) {
+			const value = this.head.value;
```
can we add the type?

bfitzpat replied:
will do types in a separate PR

lhein commented on src/doublyLinkedList.ts:
```diff
@@ -0,0 +1,83 @@
+			const value = this.head.value;
+			this.head = this.head.next;
+			if (!this.head) {
+				this.tail = undefined;
```
there is something wrong in this function with indenting and also the `else` being on a new line

lhein commented on src/doublyLinkedList.ts:
```diff
@@ -0,0 +1,83 @@
+	/**
+	* Adds an item in O(1)
+	**/
+	add(value: T) {
```
Suggestion: Given that you are using `pop` as the removal function name I would suggest to name the add function `push` or you make it the other way round and rename the `pop` to `remove` or similar.

lhein commented on src/doublyLinkedList.ts:
```diff
@@ -0,0 +1,83 @@
+	 */
+	pop(): T | undefined {
+		if (this.tail) {
+			const value = this.tail.value;
```
can you add a type?

bfitzpat replied:
will do types in a separate PR

lhein commented on src/doublyLinkedList.ts:
```diff
@@ -0,0 +1,83 @@
+	 * Returns an iterator over the values
+	 */
+	*values() {
+		let current = this.head;
```
can we add types?

bfitzpat replied:
will do types in a separate PR

bfitzpat commented on media/main.js:
```diff
@@ -79,36 +90,44 @@ function () {
 				const requirementName = json.requirementName;
 				const isAvailable = json.result;
 
+				let requirementMessage = 'Not currently available';
```
it's been that way since the beginning and I've been looking at it as the general pattern "X" and "not X" for simplicity so that when you look at the first word in the status you get a quicker indication. so I'm going to leave it. :)

bfitzpat commented:
At @paoloantinori 's suggestion, I did look a bit into using the HTML5 History API (https://developer.mozilla.org/en-US/docs/Web/API/History_API) and it's an interesting take on things.

We could definitely redirect commands to the webview and use the window.history.back() and window.history.forward() commands. This may in fact be a bit better in some respects because it would cache the actual html. I actually tried this first thing and it simply doesn't seem to work. I'm not sure how to modify it *to* work. Will chat with Riccardo to see if he has any ideas.

bfitzpat commented:
@lhein I have implemented the non-type related changes (pushing those to https://github.com/redhat-developer/vscode-didact/pull/174) so hoping you might give this another look if you get a chance?

I have also dropped a note to Riccardo (he's on PTO) so maybe he'll have some ideas on the HTML5 history aspect.

bfitzpat commented:
Rebased, cleaned up main.js to remove the html5 history bits (which weren't working). May investigate caching the html as well as the uri in the future, but for now this works as a simple history until we need something more complex. Also fixed the issue where the extension wasn't activating when a webview was persisted between sessions - added the explicit `onWebviewPanel` activation event to handle that. 

But I am considering this done for a first cut.

bfitzpat commented:
Around the HTML5 history aspect, I did chat with Riccardo about it. He offered a few links:

* https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onpopstate (API doc)
* react-router example https://github.com/ReactTraining/history/blob/ef40006b921c57ef03ae5e5ec6376072095485b1/packages/history/index.ts#L684 and https://github.com/ReactTraining/history/blob/ef40006b921c57ef03ae5e5ec6376072095485b1/packages/history/index.ts#L640-L682

But honestly I did not see an easy way to bridge the gap between the webview and this javascript coding. It was not clear how to hook up an event listener that did what needed to be done. 

I think that the simpler approach I have used for now is fine for a first iteration. We can look into making it better (i.e. caching the html) moving forward (https://github.com/redhat-developer/vscode-didact/issues/184).
